eth/stagedsync/stage_bodies.go

```diff
@@ -181,53 +222,104 @@ Loop:
 				cfg.bd.RequestSent(req, currentTime+uint64(timeout), peer)
 				d3 += time.Since(start)
 			}
+
+			loopCount++
+			if loopCount >= requestLoopCutOff {
+				break
+			}
 		}
-		start := time.Now()
-		headers, rawBodies, err := cfg.bd.GetDeliveries()
+
+		start = time.Now()
+		requestedLow, delivered, err := cfg.bd.GetDeliveries(innerTx)
 		if err != nil {
-			return err
+			return false, err
 		}
+		totalDelivered += delivered
 		d4 += time.Since(start)
 		start = time.Now()
-		cr := ChainReader{Cfg: cfg.chanConfig, Db: tx}
-		for i, header := range headers {
-			rawBody := rawBodies[i]
-			blockHeight := header.Number.Uint64()
-			// Txn & uncle roots are verified via bd.requestedMap
-			err := cfg.bd.Engine.VerifyUncles(cr, header, rawBody.Uncles)
-			if err != nil {
-				log.Error(fmt.Sprintf("[%s] Uncle verification failed", logPrefix), "number", blockHeight, "hash", header.Hash().String(), "err", err)
-				u.UnwindTo(blockHeight-1, header.Hash())
-				break Loop
+		cr := ChainReader{Cfg: cfg.chanConfig, Db: innerTx}
+
+		toProcess := cfg.bd.NextProcessingCount()
+
+		if toProcess > 0 {
+			var i uint64
+			for i = 0; i < toProcess; i++ {
+				nextBlock := requestedLow + i
+				var k [8]byte
+				binary.BigEndian.PutUint64(k[:], nextBlock)
+				body, err := innerTx.GetOne("BodiesStage", k[:])
```

Comment:
what if body is nil?

Reply:
If body is nil the JSON unmarshall fails, but it would be good to be explicit here and return early, will update the code